### PR TITLE
fix(runtime): honor native runtime selection across startup paths

### LINKED ISSUE #1753: Windows: serve ignores --llama-flavor/--device and auto-reinstalls ROCm (doctor --json also ignores the flag)
## Summary

On Windows with an AMD GPU (RX 9070 XT), `mesh-llm serve --llama-flavor <cpu|vulkan|…>` and `--device <…>` have **no effect** on which native runtime flavor gets loaded. `serve` always auto-selects (and **auto-reinstalls if missing**) the ROCm runtime, even when a flavor that can never need ROCm (`cpu`) is explicitly requested.

`--llama-flavor` and `--device` are documented global flags (`mesh-llm --help-advanced`; possible flavor values: `cpu`, `cuda`, `rocm`, `vulkan`, `metal`). Clap accepts them with no parse error. Selection still ignores them.

`mesh-llm doctor --json`'s `selected_runtime_flavor` **also ignores** `--llama-flavor`. It reports whichever flavor is already installed locally (preferring rocm > vulkan > cpu when multiple are cached) and never downloads. That makes `doctor` a bad predictor of `serve`: `serve` will additionally auto-install its preferred flavor when absent.

This is **not** #1160 (Windows Vulkan MinGW `libstdc++` / LoadLibrary 127). That is fixed on v0.76.0 on this machine.

## Steps to reproduce

1. Install MeshLLM **v0.76.0 stable** on Windows. GPU: AMD RX 9070 XT. Cache cpu + vulkan runtimes. Remove ROCm:

   ```
   mesh-llm runtime remove meshllm-native-runtime-windows-x86_64-rocm --mesh-version 0.76.0
   ```

2. Confirm `mesh-llm runtime list` shows cpu + vulkan only (no rocm).

3. Run:

   ```
   mesh-llm --debug --llama-flavor cpu serve --gguf "<absolute local .gguf>" --local-model-only --port 9337 --console 3131
   ```

4. Observe the live process modules (`ggml-hip.dll`, `hipblas.dll`, `rocblas.dll`, `amdhip64_6.dll`, …) resolve from:

   `%LOCALAPPDATA%\mesh-llm\native-runtimes\0.76.0\meshllm-native-runtime-windows-x86_64-rocm\lib\`

   not the cpu runtime.

5. `mesh-llm runtime list` afterward shows ROCm **reinstalled**.

Doctor matrix (flag has zero effect in every cell):

| rocm installed? | `--llama-flavor` | `selected_runtime_flavor` |
|---|---|---|
| yes | `vulkan` | `rocm` |
| yes | *(none)* | `rocm` |
| yes | `rocm` | `rocm` |
| yes | `cpu` | `rocm` |
| no | *(none)* | `vulkan` |
| no | `rocm` | `vulkan` (doctor does **not** auto-install) |

Related observation: `--llama-flavor vulkan --device Vulkan0` with ROCm already cached still selected ROCm, then failed model open with `InvalidArgument: unknown selected backend device: Vulkan0` (device name validated against the auto-selected backend, not the requested one).

## Expected

`--llama-flavor cpu` should never need or install ROCm. `--llama-flavor vulkan` / `--device` should select that backend. `doctor --json` should report the same selection `serve` will use, or clearly say “install-state only, flags ignored.”

## Actual

Hardware + catalog availability drive selection. CLI flags do not. The only workaround found: point `MESH_LLM_NATIVE_RUNTIME_MANIFEST_URL` at a filtered `native-runtimes.json` with the Windows ROCm artifact removed **and** `runtime remove` the cached ROCm copy. Then `serve` falls back to Vulkan.

## Environment

- mesh-llm **0.76.0** (stable, not prerelease)
- Windows 11
- AMD Radeon RX 9070 XT
- Install: official `install.ps1` (no `-PreRelease`)

## Possibly related (not duplicates)

- #1299 — `--device` silently ignored on the `--split` stage load path (same “flag does nothing” class, different path)
- #1374 — `--model` bypasses pinned-GPU resolution → `unknown selected backend device: …` (same downstream FFI error we also hit)
- #1610 / #1612 — runtime-catalog precedence / adjacent CPU bundle hiding remote GPU runtimes (same subsystem, **opposite** direction)
- #1022 (closed) — earlier RX 9070 XT Vulkan enumeration; different mechanism, already fixed

### PR DESCRIPTION
**Pull request by i386**
`--llama-flavor` was parsed into `RuntimeOptions`, but native-runtime selection did not consistently receive it. On AMD or NVIDIA hosts, an explicit CPU or Vulkan request could therefore load or install a higher-ranked ROCm or CUDA bundle. Config-only backend selection had related gaps: it required a pinned Mesh version, accepted unknown backend names, and was ignored by install/setup and embedded SDK serving.

This change applies one precedence rule across serve, doctor, runtime install/list, and setup: CLI flavor → configured selection → recommended. Configured Mesh version and Skippy ABI constraints are retained, selection-only configs track the current release, and invalid configured backend names fail validation. Embedded SDK serving now resolves the backend from the same validated config snapshot instead of hardcoding `Recommended`. The existing device selector then resolves against the requested backend, so combinations such as `--llama-flavor vulkan --device Vulkan0` validate against Vulkan.

Validation at `ffd7fba29314ab9b772442ade8f4e1f7f5c26d3c`:
- full `mesh-llm-config`, `mesh-llm-commands`, `mesh-llm-host-runtime`, and `mesh-llm` package test suites
- formatting, warning-denied clippy, default and no-default-feature checks
- repository console-output consistency and diff checks

Closes #1753

### CODE REVIEW

File: crates/mesh-llm-host-runtime/src/system/native_runtime.rs [hunk 1]
```diff
@@ -62,6 +62,26 @@ mod dynamic {
...
+                    .or(config.selection.as_deref()),
+            )?;
+            Ok(match config.mesh_version {
+                Some(mesh_version) => {
+                    Self::explicit(mesh_version, config.skippy_abi, runtime_selection)
+                }
+                None => Self {
+                    runtime_selection,
+                    ..Self::current()
+                },
```
[the remark this answers is not in the document]
Comment (i386) [reply crates/mesh-llm-host-runtime/src/system/native_runtime.rs:82]: This state is rejected before `from_config` can run. `plugin::load_config` calls the validated `parse_config_toml`, and `validate_runtime_config` emits an error whenever `mesh_version` is absent while either `skippy_abi` or `selection` is present (`crates/mesh-llm-config/src/validate.rs:196-203`). The config parser already has regression coverage for rejecting a partial native-runtime override (`crates/mesh-llm-config/src/lib.rs:171-192`). Because this private helper is only called with that validated config, the `None` branch can only receive the all-default native-runtime config; preserving an ABI-only value would support an invalid state and weaken that invariant.

**Review by Bortlesboat — COMMENTED** [review]
I traced the same path for #1753 before seeing this. One gap: `mesh-llm doctor --json` still takes `selected_runtime_flavor` from config only (`crates/mesh-llm/src/commands/doctor.rs:55` passes `native_runtime.selection`), so the doctor matrix in the issue won't change with this PR, and doctor can still disagree with what `serve --llama-flavor` loads.

The fix is small: pass `cli.llama_flavor` into `dispatch_doctor_command` and prefer it over the config selection. Happy to send it as a follow-up once this lands, or you can fold it in here if you'd rather keep #1753 to one PR.

**Comment by i386** [comment]
@Bortlesboat thanks, Ill fold it in here and we can put it in the next release

**Comment by i386** [comment]
closes https://github.com/Mesh-LLM/mesh-llm/issues/1753

**Comment by i386** [comment]
@Bortlesboat Folded this into `38f2363ad`.

Doctor dispatch now receives `cli.llama_flavor` and resolves it ahead of the configured native-runtime selection, while leaving the configured Mesh version and Skippy ABI intact. I added regressions for both CLI-over-config precedence and config fallback.

Validated with the full `mesh-llm` package suite (140 passed), warning-denied all-target clippy, the no-default-features check, rustfmt, and `git diff --check`.

**Comment by i386** [comment]
@Bortlesboat @ndizazzo review would be great please :)

File: crates/mesh-llm/src/commands/doctor.rs [hunk 2]
```diff
@@ -55,13 +56,28 @@ pub(crate) async fn dispatch_doctor_command(
...
+    llama_flavor
+        .map(|flavor| match flavor {
+            BinaryFlavor::Cpu => "cpu",
+            BinaryFlavor::Cuda => "cuda",
+            BinaryFlavor::Rocm => "rocm",
+            BinaryFlavor::Vulkan => "vulkan",
+            BinaryFlavor::Metal => "metal",
+        })
+        .or(configured_selection)
+}
```
[the remark this answers is not in the document]
Comment (i386) [reply crates/mesh-llm/src/commands/doctor.rs]: Moved the flavor-over-config precedence into `mesh-llm-commands::runtime_native`, where the doctor command policy now resolves the system `BinaryFlavor` via its canonical suffix. `crates/mesh-llm` only maps the CLI enum at the dispatch boundary. The precedence regression tests moved with the policy. Verified with `cargo fmt --all --check`, the full `mesh-llm-commands` lib suite (229 tests), and the full `mesh-llm` lib suite (50 tests) at cfc9bd932.

**Review by ndizazzo — CHANGES_REQUESTED** [review]
## Review — config-side gaps in the same class as #1753

Verified against head `ffea9b366`. The startup path this PR fixes is correct, and the two precedence tests in `system/native_runtime.rs` (`cli_llama_flavor_selects_the_current_native_runtime_backend`, `cli_llama_flavor_overrides_configured_backend_without_changing_version`) cover the right contract. Per @jamesr's call to fix this in-PR rather than defer, here are four remaining holes in the same subsystem. All four still reproduce at `ffea9b366`.

---

### 1. `selection` cannot be set in config without also pinning `mesh_version` — and this PR makes the validator stricter than the code

`crates/mesh-llm-config/src/validate.rs:200-205`:

```rust
if mesh_version.is_none() && (skippy_abi.is_some() || selection.is_some()) {
    diagnostics.push(validation_diagnostic(
        "runtime.native_runtime",
        "runtime.native_runtime override must set mesh_version when skippy_abi or selection is set",
    ));
}
```

The only way to express "always Vulkan on this box" in TOML is to also pin `runtime.native_runtime.mesh_version`, which then stops tracking releases. That is the config-side equivalent of the bug this PR fixes, and it is the one operators actually hit.

`skippy_abi` genuinely needs the coupling — an ABI string is only meaningful relative to a Mesh version. `selection` does not: it names a backend, and after this PR `from_config` already parses the selection independently of `mesh_version`. So post-merge the validator rejects a configuration the runtime now handles correctly.

**Ask:** drop `selection` from that condition, leaving `mesh_version.is_none() && skippy_abi.is_some()`. Add a validator test that `selection = "vulkan"` with no `mesh_version` produces zero diagnostics.

### 2. `selection` is never validated against known backends

`RuntimeSelection::parse` (`crates/mesh-llm-native-runtime/src/resolver.rs:172`) ends in `lower.parse()?` into `NativeRuntimeBackendKind::from_str`, whose match arm is `_ => Self::Other(normalized)` (`crates/mesh-llm-native-runtime/src/flavor.rs:184`). `from_str` can only fail on an empty string. So `selection = "vulcan"` passes `mesh-llm config validate`, matches nothing in the cache, and then drives the one-shot startup install for a backend that does not exist — the same user-visible shape as the ROCm reinstall report in #1753. `--llama-flavor` is protected here only because clap's `ValueEnum` constrains the input.

**Ask:** validate `selection` in `validate_runtime_config` by calling `RuntimeSelection::parse` and rejecting a `Backend { kind: Other(_), .. }` result, with a message listing the accepted values (`recommended`, `cpu`, `metal`, `cuda`/`cudaNN`, `rocm`, `vulkan`, `exact:<id>`, `meshllm-<id>`). Keep `Other` in the type — it is load-bearing for forward-compatible manifest parsing — and reject only at the config boundary. Test: `selection = "vulcan"` yields one diagnostic; `selection = "cuda12"` and `selection = "exact:meshllm-…"` yield none.

### 3. `--llama-flavor` is a global flag but the install and setup paths ignore it — in both directions

The flag lives on the top-level `Cli` struct (`crates/mesh-llm-cli/src/parser/commands.rs:696`), so `mesh-llm --llama-flavor vulkan runtime install` parses cleanly and is silently dropped: `dispatch_runtime_command` (`crates/mesh-llm/src/commands/runtime.rs:10`) and `dispatch_setup_command` (`crates/mesh-llm/src/commands/setup.rs:12`) never receive `cli.llama_flavor`, unlike `dispatch_doctor_command` which this PR correctly wired at `crates/mesh-llm/src/commands/mod.rs:99`.

Configured `selection` is dropped on the same paths for a second, independent reason — `native_runtime_config_selector` (`crates/mesh-llm/src/commands/runtime.rs:174`) returns `None` outright when `mesh_version` is unset, discarding `selection` with it:

```rust
Ok(match config.runtime.native_runtime.mesh_version {
    Some(mesh_version) => Some(NativeRuntimeConfigSelector { mesh_version, .. }),
    None => None,
})
```

So `runtime native list --available`, `runtime install`, `runtime prune`, and `setup` have no flavor control from either the flag or the config file.

**Ask:** thread `cli.llama_flavor` into both dispatchers and reuse the existing `native_runtime_selection(llama_flavor, configured_selection)` precedence helper (`crates/mesh-llm-commands/src/runtime_native.rs:412`) rather than adding a second rule. Make `NativeRuntimeConfigSelector.mesh_version` an `Option<String>` so a configured `selection` survives an unset version — this is the same decoupling as finding 1, and doing them together keeps the CLI and config sides consistent. Test: `--llama-flavor vulkan runtime install` selects the Vulkan artifact on a host whose recommended backend is ROCm.

### 4. Embedded/SDK serving hardcodes `Recommended`

`load_local_native_runtime_for_embedded_serving` (`crates/mesh-llm-host-runtime/src/system/native_runtime.rs:101`) passes `&RuntimeSelection::Recommended` literally — no config, no flag — and its only caller is `sdk/embedded_startup.rs:20`. On an NVIDIA host with a Vulkan driver this picks CUDA every time regardless of what the operator configured, and this PR does not touch it.

**Ask:** resolve the selection from config here the way the startup path now does, defaulting to `Recommended` when nothing is configured. If the SDK entrypoint has no config handle by design, say so and I will spin this out — but it should not stay silently hardcoded.

---

### Scope note

Findings 1+2 are the config crate; 3+4 are plumbing. If the goal is to ship 0.76.1 quickly, 1+2+3 are the ones that close the operator-visible hole and are small; 4 is separable. I'd rather see all four land here than open three followups that go stale, which is why this is request-changes rather than a comment — happy to flip to approve if you'd prefer to split.

One bookkeeping item: the "38f2363ad" head reported in chat is stale; current head is `ffea9b366`. Re-confirm CI against that before merge.

**Review by Bortlesboat — APPROVED** [review]

**Comment by i386** [comment]
@ndizazzo Addressed all four findings in `ffd7fba29314ab9b772442ade8f4e1f7f5c26d3c`:

1. `runtime.native_runtime.selection` is now independent of `mesh_version`; ABI-only overrides remain rejected. The config suite covers a selection-only config.
2. Config validation parses `RuntimeSelection` and rejects `Backend::Other` at the config boundary while keeping manifest parsing forward-compatible. Tests cover `vulcan` rejection plus `cuda12`, exact-ID, and Mesh runtime-ID acceptance.
3. The top-level `--llama-flavor` now reaches runtime native list/install and setup. These paths reuse the existing CLI flavor → configured selection helper, and selection-only config survives dispatch with an optional Mesh version. Resolver and setup tests assert Vulkan wins over configured ROCm.
4. Embedded serving reads the validated config snapshot and passes its effective selection into local native-runtime loading, defaulting to `Recommended`. Tests cover configured Vulkan and the default.

I also moved embedded runtime preparation after creation of the validated config snapshot so selection and the worker consume the same immutable config image.

Full package suites for `mesh-llm-config`, `mesh-llm-commands`, `mesh-llm-host-runtime`, and `mesh-llm` pass at that exact commit. Formatting, warning-denied clippy, default/no-default-feature checks, console-output consistency, and diff checks are clean. Replacement CI is running on the pushed head.

**Comment by i386** [comment]
Follow-up `633403c31c03fa283a39e24f7df5c0d64abfb3b8` fixes the publish-chain consistency failure introduced by the new config → native-runtime dependency: `mesh-llm-native-runtime` now precedes `mesh-llm-config` in `scripts/publish-crates.sh`. The exact failing CI command and the publish dry-run both pass at this head. Replacement CI is running.

**Review by ndizazzo — APPROVED** [review]
Re-reviewed at `633403c31`. This is the right shape now, and all four items from my earlier review are in.

1. `selection` no longer drags `mesh_version` with it (`validate.rs:200`); only `skippy_abi` keeps the coupling, and `native_runtime_override_accepts_mesh_version_with_optional_abi_and_selection` flipped from expect-error to expect-ok, which is exactly the point.
2. Unknown backends are rejected at the config boundary via `RuntimeSelection::parse` with `Other` kept in the type (`validate.rs:223`), pinned by the `vulcan` / `cuda12` / `exact:` tests.
3. `cli.llama_flavor` reaches both `runtime` and `setup` dispatch (`commands/mod.rs:66`), precedence goes through the one `native_runtime_selection` helper in `mesh-llm-commands`, and `NativeRuntimeConfigSelector.mesh_version` is `Option` so a selection-only config survives.
4. Embedded serving resolves selection from the validated snapshot with `Recommended` as the default (`sdk/embedded_startup.rs:36`), with both cases tested.

I ran `cargo test` for `mesh-llm-config`, `mesh-llm-native-runtime`, `mesh-llm-commands`, and `mesh-llm` at this head locally (all green), plus fmt and warning-denied clippy. CI is green here too.

Two notes, neither blocking: the new `mesh-llm-config -> mesh-llm-native-runtime` dependency edge is small and validation-only, but it is a new crate edge we now own (and it is why the publish order moved). And `print_configured_selector` now prints a "Using native runtime selector" block for selection-only configs, which is a small stderr change on those hosts. Good to merge.

**Comment by i386** [comment]
Review follow-up is pushed at `001f4311bf5971411aa54002a11638bf21dd027e`.

- The four items in the earlier request-changes review remain addressed at the current head: selection is independent of `mesh_version`, unknown configured backends are rejected, CLI flavor reaches runtime/setup/doctor, and embedded startup honors configured selection.
- The fresh validation edge is fixed: `exact:` and whitespace-only exact IDs now fail config validation, with regression coverage. Valid exact IDs and forward-compatible manifest parsing remain accepted.

Re-requesting review against the current head.

**Review by ndizazzo — CHANGES_REQUESTED** [review]
This is the right direction and the backend half holds up under a live check. The one thing I want changed is on the console side: the setting this PR makes reachable shows up as a free-text box.

## What I ran

I built this branch at `633403c31`, built the console bundle at release profile so there is no debug or fixture UI in play, and ran an isolated node (private `HOME`, runtime root, and ports) whose entire config was:

```toml
[runtime.native_runtime]
selection = "metal"
```

It starts, which is the point of the change, since before this the same file failed validation for a missing `mesh_version`. Unknown backends still fail closed, `selection = "vulcan"` refuses to boot with your new message. Here is Configuration -> Runtime -> Show advanced on that node:

Runtime tab, native runtime selection rendered as a text input

## Let's make the selection a real select

`Selection` being free text invites operators to type `vulcan`, `CUDA`, or `metal ` into a field whose value space is a closed set plus two prefixed forms. Your validator does catch it, but only at save or restart, and the error arrives without telling anyone what was legal.

Why it lands as text today:

- `built_in_schema/declarations.rs:108` declares the setting as `ConfigValueSchema::String`.
- `apply_built_in_control_behavior` falls through to `apply_runtime_controls_behavior` (`control_behavior.rs:72`), which has no arm for `runtime.native_runtime.*`, so nothing attaches a control behavior.
- With no `options_source` and no enum, `createSchemaControl` in `schema-control-factory.ts` ends at `textControl`.

The good news is the control we want is already built and wired, this setting just never asks for it. `ConfigOptionsSource::RuntimeNativeBackends` is served by `native_backend_state()` in `runtime_control_state_sources.rs:80`, which unions the host profile's `available_flavors` with every installed runtime's backend kind, and `schema-control-factory.ts:137` already carries its `Select backend` placeholder. Declaring the source is enough to get a native select, the same way `defaults.hardware.device` gets its GPU list.

### 1. Ask for the runtime options source

In `control_behavior/shared.rs`, next to `set_runtime_gpu_options`:

```rust
pub(super) fn set_runtime_native_backend_options(setting: &mut ConfigSettingSchema) {
    control_behavior_mut(setting).options_source = Some(ConfigOptionsSource::RuntimeNativeBackends);
}
```

Then in `control_behavior/runtime_controls.rs`:

```rust
"runtime.native_runtime.selection" => {
    set_runtime_native_backend_options(setting);
    push_non_empty_constraint(setting);
}
```

The non-empty constraint mirrors the empty-selection rule you added in `validate.rs`, so the control refuses what the validator would reject anyway.

### 2. `recommended` needs to be in the list

`RuntimeSelection::parse` accepts it and it is the documented default, but `native_backend_state()` only emits installed and host-available kinds. I would like the list to read `Recommended (auto-detect)` first, then the host backends. We can either prepend it in `native_backend_state()`, or declare the schema as `OneOf[Enum["recommended"], String]` and merge static enum values ahead of runtime options in the runtime branch of `createSchemaControl`, which currently discards `enumValues`. I lean toward the second so the shared options source stays honest for the plugin settings that also use it.

### 3. The part that can actually lose data

`exact:<id>`, `meshllm-<id>`, and `cudaNN` are legal and cannot be enumerated. Today the option list is placeholder plus runtime options, so a config carrying `exact:meshllm-native-runtime-linux-x86_64-cuda12` has no matching option, and a select that renders no match and then writes on save quietly rewrites a deliberate pin.

Let's make an unmatched stored value appear as a synthetic selected option showing the raw value with a `from config file` note, keep it selected until someone picks something else, and leave `exact:` as a config-file escape hatch. Worth a round-trip regression too: load a config with an `exact:` pin, open Runtime, change an unrelated setting, save, and assert the apply-config payload still carries the pin verbatim.

### 4. Presentation metadata

`presentation.rs` has no entry for these three paths, so they inherit fallback labels, and `Selection` on its own says nothing. Unlike its neighbours the row also shows no canonical path. Let's add entries for `runtime.native_runtime.selection` (label `Native runtime backend`, `.hint("select")`), `mesh_version`, and `skippy_abi`.

### 5. Tab copy

`ConfigurationEditorTabs.tsx:224` still says "installation and hardware selection are intentionally not presented as switchable UI controls here" while the tab renders an editable selection field. Worth updating in the same change.

### Tests I would want

- Rust: assert `runtime.native_runtime.selection` carries `options_source: runtime_native_backends`, extending `api/tests/runtime_control_state.rs`.
- UI unit: `createSchemaControl` returns `kind: 'choice'` with `presentation: 'select'` and `recommended` first.
- UI e2e: `e2e/configuration/schema-controls.spec.ts` already has `runtime_native_backends` fixtures at lines 67 and 225, easier to extend those than to add a harness.
- The `exact:` round-trip from item 3.

## Not yours, just noting it

The TOML Output tab prints `selection = "metal"` under `[runtime]` instead of `[runtime.native_runtime]`, and flattens every `runtime.activity.*` key the same way, while the Validation panel still calls the result valid.

TOML Output tab flattening nested runtime sub-tables

Saving goes through a structured merge rather than this text, so it looks display-only, but the pane has a Copy button, and copying that into `config.toml` drops the backend selection. This PR touches no UI files so it is pre-existing. I will file it separately.

**Comment by i386** [comment]
@ndizazzo addressed the console-side changes-request in `b0e59c2b0`. The native runtime backend is now a real select instead of free text:

1. **Options source + closed-set guard** — `runtime.native_runtime.selection` now declares `options_source: RuntimeNativeBackends` (`control_behavior/shared.rs` + `runtime_controls.rs`) plus a non-empty constraint that mirrors your empty-selection validator rule.
2. **Recommended first** — took your second option: the schema is now `OneOf[Enum["recommended"], String]`, and `createSchemaControl`'s runtime branch merges static enum values ahead of the runtime options instead of discarding them, so the list reads *Recommended (auto-detect)* → host backends. The shared options source stays honest (no `recommended` injected into `native_backend_state()`), so plugin settings that reuse it are unaffected.
3. **exact: pins don't lose data** — an unmatched stored value (`exact:<id>`, `meshllm-<id>`, `cudaNN`, or a GPU device that's gone) is surfaced as a synthetic trailing option labelled with the raw value and a *Set in config file* note, kept selected, so a select never renders no-match and silently rewrites the pin on save. `exact:` stays a config-file escape hatch.
4. **Presentation metadata** — added entries for `selection` (label *Native runtime backend*, `.hint("select")`), `mesh_version`, and `skippy_abi`.
5. **Tab copy** — updated `ConfigurationEditorTabs.tsx` so it no longer claims selection isn't a switchable control.

Tests, per your list:
- **Rust**: `runtime_control_state.rs` now asserts `runtime.native_runtime.selection` carries `options_source: runtime_native_backends` (and surfaces the backend options through the builder).
- **UI unit**: `createSchemaControl` returns a `choice`/`select` with `recommended` first; `resolvedChoiceOptions` appends the synthetic pin option.
- **e2e**: extended `schema-controls.spec.ts` with the `exact:` round-trip — load a config with a pin, open Runtime, edit an unrelated setting, save, and assert the apply-config payload still carries the pin verbatim.

Green locally: `mesh-llm-config` (211 tests) + `mesh-llm-host-runtime --lib` (3049) + host/config clippy and fmt; UI vitest (1681) + lint + typecheck; all 5 `schema-controls` e2e tests. The out-of-scope TOML Output flattening you noted is untouched here.

**Review by ndizazzo — APPROVED** [review]
Re-checked at `b0e59c2b0`. This is the right direction, the console side now matches what the backend accepts. I built the branch at release profile with the console bundle embedded and ran two isolated nodes (private `HOME`, runtime root, and ports, with an owner identity initialised so the config snapshot actually loads into the editor) whose config file was only `[runtime.native_runtime]` with a `selection`.

`selection = "metal"`, Configuration -> Runtime -> Show advanced:

Runtime tab, native runtime backend rendered as a select with Metal selected

The option list, expanded for the capture. Recommended (auto-detect) comes first, then the backends the options source reports for this host (CPU, Metal, Vulkan on an M4 Max), which is exactly what `/api/runtime/config-control-state` returns for `runtime.native_runtime.selection`:

Native runtime backend options expanded

`selection = "exact:meshllm-native-runtime-darwin-aarch64-metal"`: the pin shows up as the synthetic selected option with the "Set in config file" note. The round-trip holds too. I changed CPU threads and saved, the apply-config payload still carried the pin verbatim (200, `apply_mode: staged`), and the config file on disk kept it.

Native runtime backend with an exact pin

One thing worth a small follow-up, visible in that last capture: the pinned field is also painted red with "Native runtime backend must be one of: recommended, , cpu, metal, vulkan." That comes from `acceptedValuesForSetting` in `schema-control-utils.ts`, which unions the schema enum with the control options, so it picks up the empty placeholder value (the stray ", ,") and doesn't know about the synthetic stored option. It doesn't block save, but it tells an operator their deliberate pin is invalid right next to a note saying it came from the config file. Dropping the placeholder value from the accepted list and treating the current stored value as accepted would close it. Fine to land here or separately, your call.

Approving, the select does what we asked and the pin survives a save.
